### PR TITLE
Core/Scripts Script Npc needed for Quest Torch Tossing, Part 2

### LINKED ISSUE #6879: [Event Quest] Torch Tossing
Greetings,
I have encountered a bugged quest in the Midsummer festival event.
In the quest Torch Tossing you are supposed to use the item "Practice Torches" near the Midsummer Bonfire.
I am not quite sure if the bug is the quest or the item itself.
If the quest/item would have working properly, you could use the Practice Torches near the Midsummer Bonfire and get credit for using it.

Thanks in advance,
Shadowfang

Hash: c430b45
Tdb: 335.11.47+

http://old.wowhead.com/quest=11731
http://old.wowhead.com/quest=11922
http://old.wowhead.com/item=34862

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64451-event-quest-torch-tossing?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sirikfoll**
Script for Npc Torch Tossing Target_Bunny_Controller, responsable for handle the targets rotation for Midsummer Fire Festival quest, Torch Tossing.
Correct chaining, allowed classes, remove duplicates for quest Torch Tossing,
Also fixes quest A Thief's Reward.

Note:
A bunch of Npcs from Midsummer Fire Festival in general are missing in current TDB, including some needed for this quest.
So, with this PR and the 8197467c8787397ff58cd77e8a03a22b8654cac0 the quest is fully scripted, but remains correctly spawn the npcs.

Ps: If anyone knows a better way to do this, please tell.

Closes #6879

### CODE REVIEW

File: src/server/scripts/World/npcs_special.cpp [hunk 1]
```diff
@@ -519,6 +519,75 @@ class npc_dancing_flames : public CreatureScript
 };
 
 /*######
+## npc_torch_tossing_target_bunny_controller
+######*/
+
+enum TorchTossingTarget
+{
+    NPC_TORCH_TOSSING_TARGET_BUNNY      = 25535,
+    SPELL_TARGET_INDICATOR              = 45723,
```
Comment (Goatform) [inline src/server/scripts/World/npcs_special.cpp]: You don't need comma here.

**Comment by Demonid** [comment]
Nice to see someone giving love to World Events :+1:

**Comment by tkrokli** [comment]
Agreed. Thanks for taking your time with this, sirikfoll. :star:

**Comment by Kittnz** [comment]
Hmmm i got no idea if we have sniffs for correct spawns for the old world but we can sniff for northrend and outland and fix the spawns there

File: src/server/scripts/World/npcs_special.cpp [hunk 2]
```diff
@@ -519,6 +519,75 @@ class npc_dancing_flames : public CreatureScript
...
+
+                _targetTimer = 3100;
+            }
+            else
+                _targetTimer -= diff;
+        }
+
+    private:
+        uint32 _targetTimer;
+        GuidList _targetGuids;
```
Comment (jackpoz) [inline src/server/scripts/World/npcs_special.cpp]: this could be moved as local variable inside DoSearchForTargets()

Comment (sirikfoll) [inline src/server/scripts/World/npcs_special.cpp]: Is it ok if I remove the use of the GuidList? like this
```cpp
ObjectGuid DoSearchForTargets()
        {
            std::list<Creature*> targets;
            me->GetCreatureListWithEntryInGrid(targets, NPC_TORCH_TOSSING_TARGET_BUNNY, 60.0f);
            for (std::list<Creature*>::const_iterator itr = targets.begin(); itr != targets.end(); ++itr)
            {
                if ((*itr)->GetGUID() == _lastTargetGUID)
                    targets.remove(*itr);
            }

            if (!targets.empty())
            {
                _lastTargetGUID = Trinity::Containers::SelectRandomContainerElement(targets)->GetGUID();
                targets.clear();

                return _lastTargetGUID;
            }
            return ObjectGuid::Empty;
        }

File: src/server/scripts/World/npcs_special.cpp [hunk 3]
```diff
@@ -519,6 +519,75 @@ class npc_dancing_flames : public CreatureScript
...
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (_targetTimer < diff)
+            {
+                if (Unit* target = ObjectAccessor::GetUnit(*me, DoSearchForTargets()))
+                    target->CastSpell(target, SPELL_TARGET_INDICATOR, true);
+
+                _targetTimer = 3100;
```
Comment (jackpoz) [inline src/server/scripts/World/npcs_special.cpp]: any particular reason for this timer to be set to 3000 first and then 3100 ?

File: src/server/scripts/World/npcs_special.cpp [hunk 4]
```diff
@@ -519,6 +519,75 @@ class npc_dancing_flames : public CreatureScript
...
+                if ((*itr)->GetGUID() == _lastTargetGUID)
+                    continue;
+
+                targetGuids.push_back((*itr)->GetGUID());
+            }
+
+            if (!targetGuids.empty())
+            {
+                _lastTargetGUID = Trinity::Containers::SelectRandomContainerElement(targetGuids);
+                targetGuids.clear();
```
Comment (jackpoz) [inline src/server/scripts/World/npcs_special.cpp]: targetGuids will be deleted as soon as it goes out of scope, so when DoSearchForTargets() exits, so it's not needed to call .clear()

File: src/server/scripts/World/npcs_special.cpp [hunk 5]
```diff
@@ -519,6 +519,71 @@ class npc_dancing_flames : public CreatureScript
...
+        }
+
+        ObjectGuid DoSearchForTargets()
+        {
+            std::list<Creature*> targets;
+            me->GetCreatureListWithEntryInGrid(targets, NPC_TORCH_TOSSING_TARGET_BUNNY, 60.0f);
+            for (std::list<Creature*>::const_iterator itr = targets.begin(); itr != targets.end(); ++itr)
+            {
+                if ((*itr)->GetGUID() == _lastTargetGUID)
+                    targets.remove(*itr);
```
Comment (joschiwald) [inline src/server/scripts/World/npcs_special.cpp]: use list::remove_if to prevent iterator invalidation

Comment (sirikfoll) [inline src/server/scripts/World/npcs_special.cpp]: Thanks!

**Pull request merged**